osbuild: use bootc install to deploy the container

jbtrystram opened the pull request:
Instead of deploying the container to the tree then copy all the contents to the disk image, use bootc to directly manage the installation to the target filesystems.

Right now this requires to use the image as the buildroot so this requires python (for osbuild). This is tracked in [1].

[1] https://github.com/bootc-dev/bootc/issues/1410 Requires https://github.com/osbuild/osbuild/pull/2149

dustymabe commented:
A few diffs picked up by `cosa diff --metal` from https://github.com/coreos/coreos-assembler/pull/4226

[cosa-diff-metal.txt](https://github.com/user-attachments/files/21331714/cosa-diff-metal.txt)


We should probably profile each diff (maybe in https://github.com/coreos/fedora-coreos-tracker/issues/1827) and evaluate whether it's a change we want to make or not.

dustymabe commented:
I can't get a built qemu image to boot. I suspect probably the root= and boot= UUIDs added on the kernel command line?

jbtrystram commented:
Ok interesting I found that in the built with bootc image the `var` directory is not empty: 
```
$ ~ls qemu-mount/ostree/deploy/default/var                                                                                                                                                      
lib  run  tmp
```

In the initramfs, links under `/sysroot` e.g. `/sysroot/etc -> var/etc` exist but `/sysroot/var/etc` does not exist. Could that be why `adduser fail ?`
@PeaceRebel is seeing the same issue when doing `cosa import` a container build I think.

jbtrystram commented:
Blocked on https://github.com/bootc-dev/bootc/issues/1441

jbtrystram commented on src/osbuild-manifests/coreos.osbuild.x86_64.mpp.yaml:
```diff
@@ -39,11 +39,11 @@ mpp-vars:
-    then: "name:deployed-tree"
-    else:
-      mpp-format-string: '{host_as_buildroot}'
+  buildroot: "name:deployed-tree"
+    # mpp-if: osname in ['rhcos', 'scos', 'fc']
+    # then: "name:deployed-tree"
+    # else:
+    #   mpp-format-string: '{host_as_buildroot}'
```
@dustymabe is there a way to condition that to rawhide somehow ? 
I suppose we can't have a different manifest file for rawhide ?

dustymabe replied:
We'd have to use an input variable if we wanted that. We could possibly have used ostree-ref but it's actually no longer used so I opened https://github.com/coreos/coreos-assembler/pull/4241

jbtrystram commented:
follow-up : either find a way to get the boot components inside cosa, or change the bootc code to call bootupd from the deployed root . I think the latter is preferable. 
I filed https://github.com/bootc-dev/bootc/issues/1455

jbtrystram commented:
Alright, marking this as ready for review as all the bits are in place.
I guess i need to update the osbuild manifest or the other arches as well, but I'll do that after a review to reduce the amount of back and forth. 

This will need a release of bootc.

dustymabe commented on src/osbuild-manifests/coreos.osbuild.x86_64.mpp.yaml:
```diff
@@ -644,57 +472,25 @@ pipelines:
-      - type: org.osbuild.bootupd
+      # set up the `ignition.firstboot` stamp at the end because
+      # bootc want empty filesystems
+      - type: org.osbuild.ignition
         options:
-          static-configs: true
-          deployment:
-            default: true
```
Thanks for looking into it!

jbtrystram replied:
ok so I dig a bit more into it. the `--write-uuid` creates the following:
```
cat /boot/grub2/bootuuid.cfg 
set BOOT_UUID="04968af0-1e82-40e9-bbe5-a2647797d6f5"
```

which match the boot partition UUID as expected. I see ignition change the uuid [of the filesystem](https://github.com/coreos/fedora-coreos-config/blob/c36249424f7447c40a0ae0a756c84ea5a70d951d/overlay.d/05core/usr/lib/dracut/modules.d/40ignition-ostree/ignition-ostree-firstboot-uuid#L6) at first boot so that doesn´t overlap.

jlebon commented:
> Some comments.
> 
> I think there are a few things we need to iron out before we can really move forward with this:
> 
>     1. supporting both old and new paths at the same time
> 
> 
> Do we need to? Usually when we make a change this large we roll it out slowly, which means we have to support both ways for some time.
> 
> This PR is ignoring that fact, but TBH looking at OSBuild configs that support both would be pretty intimidating, so I'm not excited about trying to do that either. I'd be interested in @jlebon or @travier's thoughts.

Rolling this out in FCOS first would be preferable indeed. Instead of adding mpp conditionals, could we just have a separate set of manifests to use temporarily? Not ideal either, I know.

If we're sufficiently convinced by the diff, we could do a hard cut over. The problem is in ensuring the diff captures everything well. I think it helps a lot though that the main changes here are happening at the filesystem level.

jbtrystram commented:
requires https://github.com/coreos/bootupd/pull/990

jbtrystram commented:
Rebased and removed osbuild patches as they have landed upstream now (requires a COSA rebuild though)

dustymabe commented on src/supermin-init-prelude.sh:
```diff
@@ -73,3 +73,13 @@ touch /etc/cosa-supermin
+# have a prepare-root config in the build environnement for bootc to read.
+# This workaround can be removed when https://github.com/bootc-dev/bootc/issues/1410
+# is fixed or we have python in all streams which allows us to use the OCI image as the buildroot.
+# Note that RHCOS and SCOS use the OCI as buildroot so they should not be affected by this.
```
still kinda wish we could just get python in here too to get away from the shenanigans

dustymabe commented on src/cmd-osbuild:
```diff
@@ -405,9 +405,23 @@ main() {
     fi
-    $cmd -- /usr/lib/coreos-assembler/runvm-osbuild                                                     \
+
+    # Use the bootc install to-filesystem manifest for rawhide
+    # See discussion in https://github.com/coreos/coreos-assembler/pull/4224#pullrequestreview-3076134370
+    # Remove when we switch to use bootc install to-filesystem to all streams
+    releasever=${build%%.*}
+    with_bootc_install=""
+    if [ "$releasever" -ge 44 ]; then
+        with_bootc_install=".bootc"
+    fi
```
This is effective, but brittle. It would probably give us the most flexibility if we made it configurable via an environment variable and also possibly a manifest setting like was done over in
https://github.com/coreos/fedora-coreos-config/pull/3754 and https://github.com/coreos/coreos-assembler/commit/532225d620e57b90d7b804d57177fa4af15bb925

i.e. I could see us wanting to roll this out in `next` and also RHCOS 10/SCOS 10 first and having those knobs would help that.

dustymabe commented on src/osbuild-manifests/coreos.osbuild.x86_64.bootc.mpp.yaml:
```diff
@@ -0,0 +1,560 @@
+      # inside our disc image
+      - type: org.osbuild.bootc.install-to-filesystem
+        inputs: 
+          images:
+            type: org.osbuild.containers
+            origin: org.osbuild.pipeline
+            references:
+              name:oci-archive:
+                name: coreos.ociarchive
```
We are losing some compat here. In the existing code we have the ability to execute either via an ociarchive OR a container repo:tag

https://github.com/coreos/coreos-assembler/blob/e9034eed1341d49138878550ef7633611cce8f08/src/osbuild-manifests/coreos.osbuild.x86_64.mpp.yaml#L191-L192

I don't actually know if this is important. I think in practice we have just been executing the ociarchive flow. But custom-coreos-disk-images at one point I think allowed the repo:tag flow.

That being said, I do have a local patch set that might be a reason to use the repo:tag flow again.

jbtrystram replied:
fixed in the last push

jbtrystram commented:
Ok I cleaned up the commits and addressed everything. 
I started a full test run with a [COSA image containing this PR](https://quay.io/repository/jbtrystramtestimages/cosa?tab=tags) and the [manifest config knob](https://github.com/jbtrystram/fedora-coreos-config/commit/5a2169978a6424cfff9257bab2a6def5417eb882) in the [staging pipeline](https://jenkins-fedora-coreos-pipeline.apps.ocp.stg.fedoraproject.org/job/jbtrystram-build/6)

jbtrystram commented:
> Ok I cleaned up the commits and addressed everything. I started a full test run with a [COSA image containing this PR](https://quay.io/repository/jbtrystramtestimages/cosa?tab=tags) and the [manifest config knob](https://github.com/jbtrystram/fedora-coreos-config/commit/5a2169978a6424cfff9257bab2a6def5417eb882) in the [staging pipeline](https://jenkins-fedora-coreos-pipeline.apps.ocp.stg.fedoraproject.org/job/jbtrystram-build/6)

a few tests are failing, i'll look into it tomorrow

jbtrystram commented:
Fixed the labels issue in the last commit

jbtrystram commented:
`ext.config.boot.grub2-install` is fixed with https://github.com/bootc-dev/bootc/pull/1816 which basically replicates what we've been doing before in osbuild

dustymabe commented on src/cmdlib.sh:
```diff
@@ -774,7 +800,7 @@ rc=0
   (cd ${workdir}; bash ${tmp_builddir}/cmd.sh |& tee /dev/virtio-ports/cosa-cmdout) || rc=\$?
 else
-  (cd ${workdir}; bash)
+  (cd ${workdir}; RUNVM_SHELL=${RUNVM_SHELL:-} bash)
```
I'm on the fence about this suggestion but it may be a little more obvious what this var is used for if we came up with a new var for it:


```suggestion
  (cd ${workdir}; RUNVM_INTERACTIVE=${RUNVM_SHELL:-} bash)
```

dustymabe commented on src/cmdlib.sh:
```diff
@@ -153,28 +153,53 @@ yaml2json() {
+    # cosa initialized dir and use `src/config`.
+    # We redirect the output to /dev/null to avoid the noisy `dirs` output.
+    set +u
+    pushd "${configdir:-src/config}" > /dev/null
```
honestly, even with your comment above I'm not sure I understand why using `$configdir` is needed here. This was clearly working before (only callers being in `cmd-fetch` and `cmd-build`) and `cmd-osbuild` doesn't call `prepare_build`, which is where `$configdir` gets set, right?

dustymabe commented on src/cmdlib.sh:
```diff
@@ -745,10 +770,11 @@ runvm() {
-    cat <<EOF >> "${vmpreparedir}/hostfiles"
-/usr/lib/osbuild/stages/org.osbuild.coreos.live-artifacts.mono
-/usr/lib/osbuild/stages/org.osbuild.coreos.live-artifacts.mono.meta.json
-EOF
```
this was added in aea526e and should have been cleaned up long ago. Can you put it in a separate commit?

dustymabe commented on src/cmdlib.sh:
```diff
@@ -745,10 +770,11 @@ runvm() {
+# Include arbitrary files from the host  
+#     cat <<EOF >> "${vmpreparedir}/hostfiles"
+# /usr/lib/osbuild/stages/org.osbuild.ln
+# /usr/lib/osbuild/stages/org.osbuild.ln.meta.json
```
Can clean this up now the `ln` stage is in a release of osbuild and also included in the RPM.

dustymabe reviewed (commented):
Thanks for pushing this up hill for so long JB.

Since this PR has been open for a long time it might help to reset it by adding an initial commit that just copies the `coreos.osbuild.x86_64.mpp.yaml` to `coreos.osbuild.x86_64.bootc.mpp.yaml` and then apply your changes on top in later commits.

Eventually that initial commit would do the copy for all architectures, but for now just x86_64 would suffice.

jbtrystram commented on src/cmdlib.sh:
```diff
@@ -153,28 +153,53 @@ yaml2json() {
+    # When called very early (e.g. cmd-fetch), configdir isn't initialized yet so we assume we are in the top
+    # cosa initialized dir and use `src/config`.
+    # We redirect the output to /dev/null to avoid the noisy `dirs` output.
+    set +u
```
I don't remember :D probably a leftover. Removed

jbtrystram commented:
Ok this is now looking ready and ready for another review !

dustymabe commented:
OK did a deep dive here today. Got distracted with a few things like python :( (added a comment to [1730](https://github.com/coreos/fedora-coreos-tracker/issues/1730#issuecomment-4165225687)) and also simplifying our manifests just in general to make them more maintainable (AI actually gave me some good insights today on this).

Here's some general comments:


---

Trying to do `cosa diff --metal` I got an error:

```
2026-03-31 16:26:08,111 INFO - Running command: ['tar', '-xf', '/tmp/tmpait9uwqf.tar', '-C', 'tmp/diff-cache/metal/43.20260331.20.dev1', '--transform', 's|[[:xdigit:]]{64}|XXXXXXXXXXXXXXXX|gx']
tar: ./ostree/bootc/storage/overlay/backingFsBlockDev: Cannot mknod: Operation not permitted
```

The file inside the disk image is:

```
[core@cosa-devsh ~]$ sudo ls -l --color /ostree/bootc/storage/overlay/backingFsBlockDev                                
brw-------. 1 root root 259, 3 Mar 31 15:24 /ostree/bootc/storage/overlay/backingFsBlockDev
```

and I had to hack `cmd-diff` to get the diff to work:

```patch
diff --git a/src/cmd-diff b/src/cmd-diff
index 5424fa497..78051e544 100755
--- a/src/cmd-diff
+++ b/src/cmd-diff
@@ -568,6 +568,7 @@ def diff_metal_helper(diff_from, diff_to):
                 # in filenames with XXXXXXXXXXXXXXXX so that we can get a real diff between
                 # two of the same files in different builds.
                 runcmd(['tar', '-xf', tmp_tar.name, '-C', diff_dir,
+                        '--exclude', '*backingFsBlockDev',
                         '--transform', 's|[[:xdigit:]]{64}|XXXXXXXXXXXXXXXX|gx'])
 
         except Exception as e:
```


The most disturbing thing I see in the actual diff is the origin file has changed:


```patch
diff --git a/tmp/diff-cache/metal/43.20260309.20.dev0/ostree/deploy/fedora-coreos/deploy/XXXXXXXXXXXXXXXX.0.origin b/tmp/diff-cache/metal/43.20260309.20.dev1/ostree/deploy/fedora-coreos/deploy/XXXXXXXXXXXXXXXX.0.origin
index b1f437b..60d27fc 100644
--- a/tmp/diff-cache/metal/43.20260309.20.dev0/ostree/deploy/fedora-coreos/deploy/XXXXXXXXXXXXXXXX.0.origin
+++ b/tmp/diff-cache/metal/43.20260309.20.dev1/ostree/deploy/fedora-coreos/deploy/XXXXXXXXXXXXXXXX.0.origin
@@ -1,2 +1,2 @@
 [origin]
-container-image-reference=ostree-image-signed:docker://quay.io/fedora/fedora-coreos:testing-devel
+container-image-reference=ostree-unverified-registry:ostree-image-signed:docker://quay.io/fedora/fedora-coreos:testing-devel
```

If I'm not mistaken doesn't zincati use the origin file and this might throw it off?

travier reviewed (dismissed):
Approved but let's wait for the sync with Dusty before merging.

dustymabe commented:
Since there is a merge conflict now (with the merge of https://github.com/coreos/coreos-assembler/pull/4526) I'll rebase this.
